File: Ruby/lib/parrot.rb
class Parrot

  def initialize type, number_of_coconuts, voltage, nailed
    @type = type
    @number_of_coconuts = number_of_coconuts
    @voltage = voltage
    @nailed = nailed
  end

  def speed
    case @type
    when :african_parrot
      return [0, base_speed - load_factor * @number_of_coconuts].max
    when :norwegian_blue_parrot
      return (@nailed) ? 0 : compute_base_speed_for_voltage(@voltage)
    end

    throw "Should be unreachable!"
  end

  private

  def compute_base_speed_for_voltage voltage
    [24.0, voltage * base_speed].min
  end

  def load_factor
    9.0
  end

  protected
  
  def base_speed
    12.0
  end
end

class EuropeanParrot < Parrot

  def initialize 
  end

  def speed
    return base_speed
  end
  
end

class AfricanParrot < Parrot

  def initialize type, number_of_coconuts, voltage, nailed
    @number_of_coconuts = number_of_coconuts
  end

  def speed
    return [0, base_speed - load_factor * @number_of_coconuts].max
  end

  private

  def load_factor
    9.0
  end
   
end

File: Go/parrot.go
package parrot

import (
	"math"
)

// Parrot has a Speed.
type Parrot interface {
	Speed() (float64, error)
}

type baseParrot struct {
}

type europeanParrot struct {
	base baseParrot
}

func EuropeanParrot() europeanParrot {
	return europeanParrot{baseParrot{}}
}

type AfricanParrot struct {
	base             baseParrot
	numberOfCoconuts int
}

type NorwegianBlueParrot struct {
	base    baseParrot
	voltage float64
	nailed  bool
}

func (parrot baseParrot) baseSpeed() float64 {
	return 12.0
}

func (parrot europeanParrot) Speed() (float64, error) {
	return parrot.base.baseSpeed(), nil
}

func (parrot AfricanParrot) Speed() (float64, error) {
	return math.Max(0, parrot.base.baseSpeed()-parrot.loadFactor()*float64(parrot.numberOfCoconuts)), nil
}

func (parrot AfricanParrot) loadFactor() float64 {
	return 9.0
}

func (parrot NorwegianBlueParrot) Speed() (float64, error) {
	if parrot.nailed {
		return 0, nil
	}
	return parrot.computeBaseSpeedForVoltage(parrot.voltage), nil
}

func (parrot NorwegianBlueParrot) computeBaseSpeedForVoltage(voltage float64) float64 {
	return math.Min(24.0, voltage*parrot.base.baseSpeed())
}
